libdnf/rpm/solv_query.cpp

```diff
@@ -1191,6 +1335,239 @@ void SolvQuery::Impl::filter_reldep(Id libsolv_key, libdnf::sack::QueryCmp cmp_t
     }
 }
 
+void SolvQuery::Impl::filter_nevra(
+    const Nevra & pattern,
+    bool cmp_glob,
+    libdnf::sack::QueryCmp cmp_type,
+    solv::SolvMap & filter_result,
+    bool with_src) {
+    Pool * pool = sack->pImpl->get_pool();
+
+    auto & name = pattern.get_name();
+    const char * name_c_pattern = name.c_str();
+    auto name_cmp_type = remove_glob_when_unneeded(cmp_type, name_c_pattern, cmp_glob);
+    bool all_names = cmp_glob && (name == "*");
+
+    auto & epoch = pattern.get_epoch();
+    const char * epoch_c_pattern = epoch.c_str();
+    auto epoch_cmp_type = remove_glob_when_unneeded(cmp_type, epoch_c_pattern, cmp_glob);
+    bool all_epoch = cmp_glob && (epoch == "*");
+    bool test_epoch = !epoch.empty();
```

Comment:
Nice work. But at this place it seems complicated, confusing, and non optimal.
Why is `all_epoch` passed to `is_valid_candidate()`?
Confusing example: `all_epoch == true` -> `epoch == "*"` -> `test_epoch == true` . The result says that: I want all epochs, but at the same time I want to filter epochs.
Because `test_epoch == true` unnecessary code is run in the `is_valid_candidate()`. Wasted time.

What about this?
`test_epoch = !all_epoch && !epoch.empty();` And simplify both `is_valid_candidate()` functions. They don't need all_epoch argument.
And do the same for `all_version`, `all_release`, `all_arch` variables.

Or overlooked I something?

Reply:
Fixed